File: restChatNew.js
// Rest based chat client
// Jim Skon 2022
// Kenyon College

var baseUrl = 'http://13.58.143.75:5005';
var state="off";
var myname="";
var inthandle;
var inthandle2;
var inthandle3;
var inthandle4;
var inthandle5 = setInterval(fetchUsers, 1000);


/* Start with text input and status hidden */
document.getElementById('chatinput').style.display = 'none';
document.getElementById('status').style.display = 'none';
document.getElementById('leave').style.display = 'none';
// Action if they push the join button
document.getElementById('login-btn').addEventListener("click", (e) => {
	join();
})

/* Set up buttons */
document.getElementById('leave-btn').addEventListener("click", leaveSession);
document.getElementById('send-btn').addEventListener("click", sendText);
// Watch for enter on message box
document.getElementById('message').addEventListener("keydown", (e)=> {
    if (e.code == "Enter") {
	sendText();
    }   
});


// Call function on page exit
window.onbeforeunload = leaveSession;


function completeJoin(results) {
	var status = results['status'];
	console.log(status)
	if (status != "success") {
		alert("Account Not Found");
		leaveSession();
		return;
	}
	var user = results['user'];
	console.log("Join:"+user);
	//https://stackoverflow.com/questions/3180710/javascript-change-p-content-depending-on-select-option
	startSession(user);
}

function join() {
	myname = document.getElementById('yourname').value;
	mypass = document.getElementById('yourpass').value;
	fetch(baseUrl+'/chat/join/'+myname+'/'+mypass, {
        method: 'get'
    })
    .then (response => response.json() )
    .then (data =>completeJoin(data))
    .catch(error => {
        {alert("Error: Something went wrong:"+error);}
    })
}

function completeSend(results) {
	var status = results['status'];
	document.getElementById('message').value = '';
	if (status == "success") {
		console.log("Send succeeded")
	} else {
		alert("Error sending message!");
	}
}

//function called on submit or enter on text input
function sendText() {
    var message = document.getElementById('message').value;
	document.getElementById('message').value = '';
    console.log("Send: "+myname+":"+message);
	fetch(baseUrl+'/chat/send/'+myname+'/'+message, {
        method: 'get'
    })
    .then (response => response.json() )
    .then (data =>completeSend(data))
    .catch(error => {
        {alert("Error: Something went wrong:"+error);}
    })    
	
}



function completeFetch(result) {
	messages = result["messages"];
	messages.forEach(function (m,i) {
		name = m['user'];
		message = m['message'];
		document.getElementById('chatBox').innerHTML +=
	    	"<font color='red'>" + name + ": </font>" + message + "<br />";
	});
}

/* Check for new messaged */
function fetchMessage() {
	fetch(baseUrl+'/chat/fetch/'+myname, {
        method: 'get'
    })
    .then (response => response.json() )
    .then (data =>completeFetch(data))
    .catch(error => {
        {console.log("Server appears to be down");}
    })
}
/* Functions to set up visibility of sections of the display */
var nameHold;
function startSession(name){
    state="on";
    
    document.getElementById('yourname').value = "";
    document.getElementById('register').style.display = 'none';
    document.getElementById('user').innerHTML = "User: " + name;
	nameHold = name;
	console.log(nameHold);
    document.getElementById('chatinput').style.display = 'block';
    document.getElementById('status').style.display = 'block';
    document.getElementById('leave').style.display = 'block';
    /* Check for messages every 500 ms */
    inthandle=setInterval(fetchMessage,500);
	inthandle2=setInterval(getUsers,500);
	inthandle3=setInterval(checkTyping,200);
	inthandle4=setInterval(updateShowTyping,200);
	inthandle5 = setInterval(fetchUsers, 500)
}

function leaveSession(){
    state="off";
    removeUser(nameHold);
    document.getElementById('yourname').value = "";
    document.getElementById('register').style.display = 'block';
    document.getElementById('user').innerHTML = "";
    document.getElementById('chatinput').style.display = 'none';
    document.getElementById('status').style.display = 'none';
    document.getElementById('leave').style.display = 'none';
	clearInterval(inthandle);
	clearInterval(inthandle2);
	clearInterval(inthandle3);
	clearInterval(inthandle4);
	clearInterval(inthandle5);

}

//FUNCTIONS THAT I HAVE ADDED BELOW-------------------------------------------------------------------





//Functions to dynamically update user list
function getUsers() {
	fetch(baseUrl+'/chat/userlist', {
        method: 'get'
    })
    .then (response => response.json() )
    .then (data =>updateUsers(data))
    .catch(error => {
        {alert("Error: Something went wrong:"+error);}
    })
}
function updateUsers(result) {
	userList = result["userList"];
	//console.log("user list printed");
	document.getElementById('userlist').innerHTML = userList;
}

//functions to register a user , /chat/register/username/email/password
document.getElementById('submitButton').addEventListener("click", registerUser);
function registerUser(){
	console.log("registerUser() running");
	username = document.getElementById('user-name').value;
	email = document.getElementById('user-email').value;
	pass = document.getElementById('user-password').value;
	fetch(baseUrl+'/chat/register/'+username +'/'+email+'/'+pass, {
        method: 'get'
    })
    .then (response => response.json() )
    .then (data =>completeRegisterUser(data))
    .catch(error => {
        {alert("Error: Something went wrong:"+error);}
    })
}

function completeRegisterUser(results){
	var status = results['status'];
	console.log(status)
	if (status != "success") {
		alert("Username or Email already exists! Password must be more than 6 characters");
		leaveSession();
		return;
	}
	var user = results['user'];
	alert("Registration Successful");
	console.log("Registered:"+user);
	username = document.getElementById('user-name').value = '';
	email = document.getElementById('user-email').value = '';
	pass = document.getElementById('user-password').value = '';
}

//Function to remove a user after they leave the site.
function removeUser(){
		fetch(baseUrl+'/chat/userlist/remove/'+nameHold, {
        method: 'get'
    })
}	

let gateway = 0;
//Functions to update the server as to whether someone is typing
function checkTyping(){
	var message = document.getElementById('message').value;
	//console.log(message);
		if(message != ""){
		updateTyping();
	} else{
		removeTyping();
	}
}

function updateTyping(){
		fetch(baseUrl+'/chat/typing/update/'+nameHold, {
        method: 'get'
    })
	gateway += 1;
	//console.log(gateway);
	//console.log("User is Typing");
}

function removeTyping(){
		gateway = 0;
		fetch(baseUrl+'/chat/typing/remove/'+nameHold, {
        method: 'get'
		})
		//console.log(gateway);
	//console.log("User is not typing");
}


function updateShowTyping(){
	if(gateway == 1){
		fetch(baseUrl+'/chat/typingmessage/'+nameHold, {
        method: 'get'
    })
		console.log(gateway);
	}
}

//IMAGE DRAG AND DROP---------------------------------------
function dragOverHandler(ev) {
  console.log("File(s) in drop zone");

  // Prevent default behavior (Prevent file from being opened)
  ev.preventDefault();
}

function dropHandler(ev) {
  console.log("File(s) dropped");

  // Prevent default behavior (Prevent file from being opened)
  ev.preventDefault();

  if (ev.dataTransfer.items) {
    // Use DataTransferItemList interface to access the file(s)
    [...ev.dataTransfer.items].forEach((item, i) => {
      // If dropped items aren't files, reject them
      if (item.kind === "file") {
        const file = item.getAsFile();
        console.log(`… file[${i}].name = ${file.name}`);
      }
    });
  } else {
    // Use DataTransfer interface to access the file(s)
    [...ev.dataTransfer.files].forEach((file, i) => {
      console.log(`… file[${i}].name = ${file.name}`);
    });
  }
}

function handleDrop(ev) {
  ev.preventDefault();
  let dt = ev.dataTransfer;
  let files = dt.files;
  handleFiles(files);
}

function handleFiles(files) {
  ([...files]).forEach(uploadFile)
}

function uploadFile(file) {
  let url = 'http://13.58.143.75:5005/chat/image/send';
  let formData = new FormData()
  formData.append('file', file)
  fetch(url, {
    method: 'POST',
	body: formData
  })
  .then(() => { /* Done. Inform the user */ })
  .catch(() => { /* Error. Inform the user */ })
}

//SEND AN EMAIL--------------
//https://kaustubh72.medium.com/send-e-mails-with-smtp-js-a8e07e1d0b6b
//https://www.smtpjs.com/

document.getElementById('sendEmailButton').addEventListener("click", sendEmail);
function sendEmail(){
let url = "http://13.58.143.75/restChat/restChat.html";
let email = document.getElementById('friend-email').value;
let body = document.getElementById('email-message').value + " " + url;
console.log(email);
Email.send({
    Host : "smtp.elasticemail.com",
    Username : "kenyonsoftwaredev@gmail.com",
    Password : "375F1C624A4AA040B8B4E8BF48D81CD73E83",
    To : email,
    From : "kenyonsoftwaredev@gmail.com",
    Subject : document.getElementById('email-subject').value,
    Body : body
})

}

//Sammy//

// NEW 
function updateUsersSam(data){
	//console.log(data);
	fetchTypers();
	const users = data['userList'];
	//Returning list like Grant,Sammy,Joe,etc. and then turning it into a javascript array.
	//https://dev.to/sanchithasr/6-ways-to-convert-a-string-to-an-array-in-javascript-1cjg
	//https://www.educative.io/answers/how-to-add-an-id-to-element-in-javascript
	const usersArray = users.split(',');
	console.log(users);
	console.log(typerArray);
	//console.log(usersArray);
    // Get the user list container element
    const userBar = document.getElementById('bottomPageList');
    // Clear any existing user list items
    userBar.innerHTML = '';
	let numGate = 0;
	let listNum = "a";
    // Create a new list item for each user and append it to the user list, first user added will have blue background
    usersArray.forEach((user) => {
		if(typerArray == null){
			if(numGate == 0){
				const listItem = document.createElement('li');
				//Set the Class of the list , then after set the ID and Text
				listItem.classList.add('list-group-item');
				listItem.classList.add('active');
				listItem.setAttribute('id' , listNum);
				listItem.textContent = user;
				userBar.appendChild(listItem);
				const spanItem = document.createElement('span');
				//Set the Class of the Span in the list , then after set the Text
				spanItem.classList.add('badge');
				spanItem.classList.add('rounded-pill');
				spanItem.classList.add('bg-success');
				spanItem.textContent = "Online";
				listItem.appendChild(spanItem);
				numGate += 1;
				listNum += "a";
			}else{
				const listItem = document.createElement('li');
				listItem.classList.add('list-group-item');
				listItem.setAttribute('id' , listNum);
				listItem.textContent = user;
				userBar.appendChild(listItem);
				const spanItem = document.createElement('span');
				spanItem.classList.add('badge');
				spanItem.classList.add('rounded-pill');
				spanItem.classList.add('bg-success');
				spanItem.textContent = "Online";
				listItem.appendChild(spanItem);
				listNum += "a";
			}
		}else{
			if(numGate == 0 && typerArray.includes(user) == true){
				let userStatus = "Is typing...";
				UpdateUsersSamTyping(user , listNum , userBar , userStatus);
				numGate += 1;
				listNum += "a";
		}else{
			if(numGate == 0 && typerArray.includes(user) == false){
				let userStatus = "Online";
				UpdateUsersSamTyping(user , listNum , userBar , userStatus);
				numGate += 1;
				listNum += "a";
		}else{
			if(numGate == 0 && typerArray.includes(user) == false){
				let userStatus = "Is typing...";
				UpdateUsersSamTyping(user , listNum , userBar , userStatus);
				numGate += 1;
				listNum += "a";
			}else{
				if(numGate != 0 && typerArray.includes(user) == true){
					let userStatus = "Is typing...";
					UpdateUsersSamNotTyping(user , listNum , userBar , userStatus);
					listNum += "a";
				}else{
					let userStatus = "Online";
					UpdateUsersSamNotTyping(user , listNum , userBar , userStatus);
					listNum += "a";
				}
			}
		}		
	}
}})};

function UpdateUsersSamTyping(user , listNum , userBar , userStatus){
			const listItem = document.createElement('li');
			//Set the Class of the list , then after set the ID and Text
			listItem.classList.add('list-group-item');
			listItem.classList.add('active');
			listItem.setAttribute('id' , listNum);
			listItem.textContent = user;
			userBar.appendChild(listItem);
			const spanItem = document.createElement('span');
			//Set the Class of the Span in the list , then after set the Text
			if(userStatus == "Is typing..."){
				spanItem.classList.add('badge');
				spanItem.classList.add('rounded-pill');
				spanItem.classList.add('bg-warning');
				spanItem.classList.add('text-dark');					
			}else{
				spanItem.classList.add('badge');
				spanItem.classList.add('rounded-pill');
				spanItem.classList.add('bg-success');	
			}
			spanItem.textContent = userStatus;
			listItem.appendChild(spanItem);
};

function UpdateUsersSamNotTyping(user , listNum , userBar , userStatus){
			const listItem = document.createElement('li');
			listItem.classList.add('list-group-item');
			listItem.setAttribute('id' , listNum);
			listItem.textContent = user;
			userBar.appendChild(listItem);
			const spanItem = document.createElement('span');
			if(userStatus == "Is typing..."){
				spanItem.classList.add('badge');
				spanItem.classList.add('rounded-pill');
				spanItem.classList.add('bg-warning');
				spanItem.classList.add('text-dark');					
			}else{
				spanItem.classList.add('badge');
				spanItem.classList.add('rounded-pill');
				spanItem.classList.add('bg-success');	
			}
			spanItem.textContent = userStatus;
			listItem.appendChild(spanItem);
};


function fetchUsers() {
    fetch(baseUrl+'/chat/users', {
        method: 'get'
    })
    .then(response => response.json())
    .then(data => updateUsersSam(data))
    .catch(error => {
        console.log("Error fetching user list:", error);
    });
}
// Call fetchUsers every 5 seconds to update the user list

function fetchTypers() {
    fetch(baseUrl+'/chat/users/typing', {
        method: 'get'
    })
    .then(response => response.json())
    .then(data => makeTyperArray(data))
    .catch(error => {
        console.log("No users are typing so there is no JSON object to fetch", error);
		typerArray = null;
    });
}

var typerArray = "";
function makeTyperArray(data){
	console.log(data);
	const users = data['typerList'];
	//Returning list like Grant,Sammy,Joe,etc. and then turning it into a javascript array.
	//https://dev.to/sanchithasr/6-ways-to-convert-a-string-to-an-array-in-javascript-1cjg
	typerArray = users.split(',');
}

File: restChatNew.cpp
#include <iostream>
#include <fstream>
#include <map>
#include <algorithm>
#include <string>
#include "httplib.h"
#include <vector>

using namespace httplib;
using namespace std;

const int port = 5005;


//Grab List of Usernames
string getUserList(map<string, string> const &activeUsers){
	string usernameList = "Users: ";
    for (auto const &pair: activeUsers) {
		usernameList += pair.first;
		usernameList += ", ";
    }
	usernameList.erase(usernameList.length() - 2);
	string jsonMessage = "{\"userList\":\""+usernameList+"\"}";
	return jsonMessage;
}

string getUserListMod(map<string, string> const &activeUsers){
	string usernameList;
    for (auto const &pair: activeUsers) {
		usernameList += pair.first;
		usernameList += ",";
    }
	usernameList.erase(usernameList.length() - 1);
	string jsonMessage = "{\"userList\":\""+usernameList+"\"}";
	return jsonMessage;
}

//Update users status if they are away or online -Sammy
//void updateStatus(map<string, string> const &holdList, map<string, string> const &activeUsers) {
//	vector<string> userList;
//	string nameHold;
//    for (auto const &pair: activeUsers) {
        // Check if user is currently active
//		nameHold = pair.first;
//        if (5 == 5) {
            // User is active, so set status to "Online"
 //           std::cout << nameHold << " is online." << std::endl;
//			userList.push_back(nameHold);
            // Change user's status to "Online" (or update existing status)
            // ...
 //       } else {
            // User is active, so set status to "Active"
//            std::cout << nameHold << " is away." << std::endl;
            // Change user's status to "Away" (or update existing status)
            // ...
 //       }
//    }
//}

//Remove someone from the active users list
void removeUser(map<string, string> &activeUsers , string username){
	activeUsers.erase(username);
}


//Add a message to a user.
void addMessage(string username, string message, map<string,vector<string>> &messageMap) {
	/* iterate through users adding message to each */
	string jsonMessage = "{\"user\":\""+username+"\",\"message\":\""+message+"\"}";
	for (auto userMessagePair : messageMap) {
		username = userMessagePair.first;
		messageMap[username].push_back(jsonMessage);
	}
}

//Check if someone is typing and then show that they are.
void showTyping(string currentUser , map<string,vector<string>> &messageMap , map<string, string> const &typingMap , map<string,string> const &isTypingMap){
	cout << "Show Typing is Running" << currentUser << endl;
	if(typingMap.count(currentUser)){
		string message = "...";
		addMessage(currentUser , message , messageMap);
	}
}

//Show ... in the chat for people typing.
void getTypersList(map<string, string> const &typingMap , map<string,vector<string>> &messageMap , map<string,string> const &isTypingMap){
	string currentUser;
    for (auto const &pair: typingMap) {
		currentUser = pair.first;
		showTyping(currentUser , messageMap , typingMap , isTypingMap);
    }
}
//Return list of people typing
string getTypersListMod(map<string, string> const &typingMap){
	string typerList;
    for (auto const &pair: typingMap) {
		typerList += pair.first;
		typerList += ",";
    }
	typerList.erase(typerList.length() - 1);
	string jsonMessage = "{\"typerList\":\""+typerList+"\"}";
	return jsonMessage;
}



//This function will add a user to a userMap, with their username, email, and password in a json string.
void addUser(string username, string password, string email, map<string,string> &userMap) {
	/* iterate through users adding message to each */
	string jsonMessage = "{\"user\":\""+username+"\",\"pass\":\""+password+"\",\"email\":\""+email+"\"}";
	userMap[username] = jsonMessage;
	cout << "addUser output: "<< userMap[username] << endl;
}

//Add someone to the active typers map
void addTyper(string username, map<string,string> &typingMap , map<string,string> &isTypingMap){
	string response = "yes";
	string message = "...";
	isTypingMap[username] = ".";
	typingMap[username] = "{\"user\":\""+username+"\",\"typing\":\""+response+"\"}";
}

//Remove someone from the active typer map.
void removeTyper(string username, map<string,string> &typingMap , map<string,string> &isTypingMap){
	string message = "";
	typingMap.erase(username);
	isTypingMap[username] = "{\"user\":\""+username+"\",\"message\":\""+message+"\"}";
}



string getMessagesJSON(string username, map<string,vector<string>> &messageMap) {
	/* retrieve json list of messages for this user */
	bool first = true;
	string result = "{\"messages\":[";
	for (string message :  messageMap[username]) {
		if (not first) result += ",";
		result += message;
		first = false;
	}
	result += "]}";
	messageMap[username].clear();
	return result;
}


int main(void) {
  Server svr;
  int nextUser=0;
  map<string,vector<string>> messageMap;
  map<string,vector<string>> typingUsersMap;
  map<string,vector<string>> holdList;
  map<string,string> userMap;
  map<string,string> userEmail;
  map<string,string> activeUsers;
  map<string,string> typingMap;
  map<string,string> isTypingMap;

	
  /* "/" just returnsAPI name */
  //This is the API home page
  svr.Get("/", [](const Request & /*req*/, Response &res) {
    res.set_header("Access-Control-Allow-Origin","*");
    res.set_content("Chat API", "text/plain");
  });

  //This is a test page
  svr.Get("/secretpage", [](const Request & /*req*/, Response &res) {
    res.set_header("Access-Control-Allow-Origin","*");
    res.set_content("You found the secret page 0_0", "text/plain");
  });
  
  //REGISTRATION: This Section should handle someone registering, /chat/register/username/email/password
  svr.Get(R"(/chat/register/(.*)/(.*)/(.*))", [&](const Request& req, Response& res) {
	res.set_header("Access-Control-Allow-Origin","*");
    string username = req.matches[1];
	string email = req.matches[2];
	string password = req.matches[3];
    string result;
    vector<string> empty;
    // Check if user with this name or email already exists, or if password is less than 6 characters.
    if (messageMap.count(username) or messageMap.count(email) or password.length() < 7){
    	result = "{\"status\":\"registrationfailure\"}";
    } else {
    	// Add users to various maps
    	messageMap[username]= empty;
		userEmail[username] = email;
		addUser(username , password, email , userMap);
    	result = "{\"status\":\"success\",\"user\":\"" + username + "\",\"email\":\"" + email + "\",\"pass\":\"" + password + "\"}";
		//Output some stuff
		cout << username << " registered" << endl;
		cout << "User Email: " << userEmail[username] << endl;

    }
    res.set_content(result, "text/json");
  });
  
  //This Section is the part of the API for logging in.
  svr.Get(R"(/chat/join/(.*)/(.*))", [&](const Request& req, Response& res) {
    res.set_header("Access-Control-Allow-Origin","*");
	//cout<< "this is log in output" << endl;
    string username = req.matches[1];
	string password = req.matches[2];
	string email = userEmail[username];
	string combined = "{\"user\":\""+username+"\",\"pass\":\""+password+"\",\"email\":\""+email+"\"}";
	//cout<< combined << endl;
	//cout<< testAgainst << endl;
    string result;
    // Check if user with this name and password exists
    if (combined == userMap[username]){
    	result = "{\"status\":\"success\",\"user\":\"" + username + "\"}";
		activeUsers[username] = "this user is active";
		cout << username << " joins" << endl;
    } else {
    	result = "{\"status\":\"failure\"}";
    }
    res.set_content(result, "text/json");
	getUserList(userMap);
  });

  //This is the part of the API that handles sending messages.
   svr.Get(R"(/chat/send/(.*)/(.*))", [&](const Request& req, Response& res) {
    res.set_header("Access-Control-Allow-Origin","*");
	string username = req.matches[1];
	string message = req.matches[2];
	string result; 
    if (!messageMap.count(username)) {
    	result = "{\"status\":\"baduser\"}";
	} else {
		addMessage(username,message,messageMap);
		result = "{\"status\":\"success\"}";
	}
    res.set_content(result, "text/json");
  });
  
 
  //This part of the code grabs messages that are sent
   svr.Get(R"(/chat/fetch/(.*))", [&](const Request& req, Response& res) {
    string username = req.matches[1];
    res.set_header("Access-Control-Allow-Origin","*");
    string resultJSON = getMessagesJSON(username,messageMap);
    res.set_content(resultJSON, "text/json");
  });
  
    //This part of the code grabs a list of users
   svr.Get(R"(/chat/userlist)", [&](const Request& req, Response& res) {
    res.set_header("Access-Control-Allow-Origin","*");
	string result;
	result = getUserList(activeUsers);
	res.set_content(result, "text/json");
  });
  
   //This part of the code will remove a user from the active user list.
   svr.Get(R"(/chat/userlist/remove/(.*))", [&](const Request& req, Response& res) {
    res.set_header("Access-Control-Allow-Origin","*");
	string username = req.matches[1];
	removeUser(activeUsers , username);
  });
  
  //This part of the code will update whether someone is typing.
     svr.Get(R"(/chat/typing/update/(.*))", [&](const Request& req, Response& res) {
    res.set_header("Access-Control-Allow-Origin","*");
	string username = req.matches[1];
	addTyper(username , typingMap , isTypingMap);
  });
  
   //This part of the code will remove someone from the typing map who is not typing.
     svr.Get(R"(/chat/typing/remove/(.*))", [&](const Request& req, Response& res) {
    res.set_header("Access-Control-Allow-Origin","*");
	string username = req.matches[1];
	removeTyper(username , typingMap , isTypingMap);
  });
 
    //This part of the code will return JSON of whether someone is typing
     svr.Get(R"(/chat/typing/(.*))", [&](const Request& req, Response& res) {
    res.set_header("Access-Control-Allow-Origin","*");
	string username = req.matches[1];
	string result = typingMap[username];
	res.set_content(result, "text/json");
  });
  
    //This part of the code will send ... for someone typing
    svr.Get(R"(/chat/typingmessage/(.*))", [&](const Request& req, Response& res) {
    res.set_header("Access-Control-Allow-Origin","*");
	string username = req.matches[1];
	string result = isTypingMap[username];
	showTyping(username , messageMap , typingMap , isTypingMap);
	res.set_content(result, "text/json");
  });
  
      //This part of the code will update the user list for the thing at the bottom of the web page
    svr.Get(R"(/chat/users)", [&](const Request& req, Response& res) {
    res.set_header("Access-Control-Allow-Origin","*");
	string result;
	result = getUserListMod(activeUsers);
	res.set_content(result, "text/json");
  });
  
    //This part of the code will update the user list for the thing at the bottom of the web page
    svr.Get(R"(/chat/users/typing)", [&](const Request& req, Response& res) {
    res.set_header("Access-Control-Allow-Origin","*");
	string result;
	result = getTypersListMod(typingMap);
	res.set_content(result, "text/json");
  });
  

  
  //What comes out in the Linux Console:
  cout << "Server listening on port " << port << endl;
  cout << "Chat Time" << endl;
  svr.listen("0.0.0.0", port);

}
